src/vcf/assembly_checker.cpp

```diff
@@ -0,0 +1,123 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "vcf/assembly_checker.hpp"
+
+namespace ebi
+{
+  namespace vcf
+  {
+    namespace assembly_checker
+    {
+      bool check_vcf_ref(std::string vcf_path, std::string fasta_path, std::string fasta_index_path)
+      {
+          std::string file_error_msg;
+
+          std::ifstream vcf_input{vcf_path};
+          file_error_msg = "Couldn't open VCF file " + vcf_path;
+          ebi::vcf::assembly_checker::check_file_validity(vcf_input, file_error_msg);
+
+          std::ifstream fasta_input{fasta_path, std::ios::binary};
+          file_error_msg = "Couldn't open FASTA file " + fasta_path;
+          ebi::vcf::assembly_checker::check_file_validity(fasta_input, file_error_msg);
+      
+          std::ifstream fasta_index_input{fasta_index_path, std::ios::binary};
+          file_error_msg = "Couldn't open FASTA index file " + fasta_index_path + ". Please use samtools "
+                                       "faidx <fasta> to create the index file";
+          ebi::vcf::assembly_checker::check_file_validity(fasta_index_input, file_error_msg);
+
+          std::vector<char> vector_line;
+          vector_line.reserve(default_line_buffer_size);
+
+          std::set<std::string> absent_chromosomes;
+          MatchStats match_stats;
+
+          // Reading FASTA index, and querying FASTA file
+          auto index = bioio::read_fasta_index(fasta_index_input);
+
+          // Reading VCF file
+          while (util::readline(vcf_input, vector_line).size() != 0) {
+              std::string line{vector_line.begin(), vector_line.end()};
+
+              if (boost::starts_with(line, "#")) {
+                  continue;
+              }
+
+              VcfVariant vcf_variant{line};
+
+              if (index.count(vcf_variant.chromosome) == 0) {
+                  absent_chromosomes.insert(vcf_variant.chromosome);
+                  continue;
+              }
+
+              if (vcf_variant.position == 0) {
+                  BOOST_LOG_TRIVIAL(warning) << "Position 0 should only be used for a telomere";
+                  continue;
+              }
+
+              auto fasta_sequence = bioio::read_fasta_contig(fasta_input, index.at(vcf_variant.chromosome),
+                                                       vcf_variant.position - 1, vcf_variant.reference_allele.length());
+              auto reference_sequence = vcf_variant.reference_allele;
+
+              match_stats.add_match_result(is_matching_sequence(fasta_sequence, reference_sequence));
+          }
+
+          BOOST_LOG_TRIVIAL(info) << "Number of matches: " << match_stats.num_matches << "/" << match_stats.num_variants;
```

Comment:
Instead of writing the report here do it from the main method. That will make this one be focused just on checking that the reference is correct, not doing UI-related tasks.

You can check the following existing code:
* [This method](https://github.com/EBIvariation/vcf-validator/blob/master/src/validator_main.cpp#L112-L150) for how to select a report type. We will only have the `MatchStats` for now, somewhat close to the error summary report in the validator. An ODB report will be added later on.
* [Entry point method](https://github.com/EBIvariation/vcf-validator/blob/master/src/vcf/validator.cpp#L133) for the validation. It can give you ideas on how to split pre-validation checks and also how to write the report. Try to follow this structure as closely as possible to improve consistency of the code across tools.
* [This class](https://github.com/EBIvariation/vcf-validator/blob/master/inc/vcf/summary_report_writer.hpp) containing the `SummaryReportWriter` mentioned above.